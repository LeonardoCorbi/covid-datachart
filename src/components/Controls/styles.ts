import styled from 'styled-components';

export const Container = styled.form`
  width: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
  gap: 32px;
  flex-wrap: wrap;
  margin-top: 40px;

  section {
    &:first-child {
      width: 50%;

      @media(max-width: 845px){
        width: 90%;
      }

      article:first-child {
        min-width: 80%;
        display: flex;
        justify-content: space-between;
        font-family: sans-serif;
      }

      article:nth-child(2) {
        ul {
          min-width: 80%;
          display: flex;
          justify-content: space-between;
          font-family: sans-serif;
          list-style: none;
          padding: 0;
        }
      }
    }

    &:nth-child(2) {
      display: flex;
      gap: 16px;
      font-family: sans-serif;

      article {
        display: flex;
        flex-direction: column;
        gap: 8px;

        @media(max-width: 460px){
          label {
            font-size: 14px;
          }
        }
      }
    }  

    input[type="range"] {
      min-width: 80%;
      position: relative;
    }
    
    [class*=PrivateValueLabel-circle] {
      transform: translateX(-24px);
      width: max-content;
      padding: 0 8px;
      border-radius: 4px;
    }
  
    [class*=PrivateValueLabel-label] {
      transform: rotate(0deg)
    }
  
    
  }
`;

export const Button = styled.button`
  background-color: #364BA8;
  color: white;
  border: none;
  border-radius: 4px;
  padding: 0 8px;
  height: 32px;
  cursor: pointer;
`;