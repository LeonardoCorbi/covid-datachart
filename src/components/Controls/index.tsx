import { ChangeEvent, Dispatch, SetStateAction, useCallback, useEffect, useMemo, useState } from 'react';
import { Slider, Select, MenuItem } from '@material-ui/core';
import { Button, Container } from './styles';
import { DATES } from '../../constants';
import { toBrazilDate } from '../../utils/date';
import { MethodType } from '../../services/Cases/dtos/requests/findByDate.request';
import { IQueryParams } from '../../../pages';

interface IControllers {
  method: string;
  selectedDate: string;
  setSelectedDate: Dispatch<SetStateAction<string>>;
  setMethod: Dispatch<SetStateAction<MethodType>>;
  updateQuery: ({ date, method: type }: Partial<IQueryParams>) => void
}

const Controls = ({
  method,
  setMethod,
  updateQuery,
  selectedDate,
  setSelectedDate,
}: IControllers) => {
  let debouncing: NodeJS.Timeout;

  const initialIndex = useMemo(() => DATES.findIndex((date) => date === selectedDate), [selectedDate]);
  const DATE_INDEX_LENGTH = DATES.length - 1;
  const TIME = 1500;

  const [sliderPosition, setSliderPosition] = useState(initialIndex);
  const [isPlayed, setIsPlayed] = useState(false);

  const onSliderChange = useCallback((
    _: any,
    value: number | number[]
  ) => {
    const date = DATES[+value];
    setSliderPosition(+value);
    clearTimeout(debouncing);
    debouncing = setTimeout(() => {
      setSelectedDate(date);
      updateQuery({ date });
    }, 500);
  }, [setSelectedDate, updateQuery]);

  const onSelectChange = useCallback((
    { target: { value } }: ChangeEvent<{ value: unknown; }>,
    _: any
  ) => {
    const type = value as MethodType;
    updateQuery({ method: type });
    setMethod(type);
  }, [updateQuery]);

  const onPlayClick = useCallback(() => {
    setIsPlayed((prevState) => !prevState);
  }, [isPlayed]);

  useEffect(() => {
    if (!(sliderPosition < DATE_INDEX_LENGTH && isPlayed)) return;
    
    const timer = setTimeout(() => {
      onSliderChange(null, sliderPosition + 1);
    }, TIME);
    
    if (sliderPosition === DATE_INDEX_LENGTH) setIsPlayed(false);
    return () => clearTimeout(timer);
  }, [sliderPosition, isPlayed]);

  useEffect(() => {
    setSliderPosition(initialIndex);
  }, [initialIndex]);

  return (
    <Container>
      <section>
        <article>
          <Slider
            min={0}
            value={sliderPosition}
            max={DATE_INDEX_LENGTH}
            valueLabelDisplay="on"
            onChange={onSliderChange}
            valueLabelFormat={(value) => toBrazilDate(DATES[value])}
          />
        </article>
        <article>
          <ul>
            <li>mai/2020</li>
            <li>out/2020</li>
            <li>mar/2021</li>
            <li>jul/2021</li>
            <li>jan/2022</li>
          </ul>
        </article>
      </section>
      <section>
        <article>
          <label htmlFor="methodSelector">Modo de visualização:</label>
          <Select
            value={method}
            defaultValue="partial"
            id="methodSelector"
            onChange={onSelectChange}
          >
            <MenuItem value="partial">Parcial</MenuItem>
            <MenuItem value="total">Total</MenuItem>
          </Select>
        </article>
        <article>
          <label htmlFor="playButton">Simular passagem de tempo:</label>
          <Button onClick={onPlayClick} id="playButton" type="button">{isPlayed ? 'Pause' : 'Play'}</Button>
        </article>
      </section>
    </Container>
  );
};

export default Controls;
