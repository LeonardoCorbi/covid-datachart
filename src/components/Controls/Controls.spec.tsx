import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import Controls, { IControls } from '.';

const component = ({
  method = 'partial',
  selectedDate = '2020-05-11',
}: Partial<IControls>) => {
  const cb = jest.fn();
  return (
    <Controls
      method={method}
      setMethod={cb}
      updateQuery={cb}
      selectedDate={selectedDate}
      setSelectedDate={cb}
    />
  );
};

describe('Control component', () => {
  it('should render correctly', () => {
    const { getByTestId } = render(component({}));

    const tree = renderer.create(component({})).toJSON();
    expect(tree).toMatchSnapshot();

    const slider = getByTestId('slider');
    const select = getByTestId('methodSelector');
    const playButton = getByTestId('playButton');

    expect(slider).toBeDefined();
    expect(select).toBeDefined();
    expect(playButton).toBeDefined();
  });
});