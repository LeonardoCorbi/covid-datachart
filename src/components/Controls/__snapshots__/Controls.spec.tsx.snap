// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Control component should render correctly 1`] = `
<form
  className="styles__Container-sc-1n80o5l-0 kBQGVK"
>
  <section>
    <article>
      <span
        className="MuiSlider-root MuiSlider-colorPrimary"
        data-testid="slider"
        onMouseDown={[Function]}
      >
        <span
          className="MuiSlider-rail"
        />
        <span
          className="MuiSlider-track"
          style={
            Object {
              "left": "0%",
              "width": "0%",
            }
          }
        />
        <input
          type="hidden"
          value="0"
        />
        <span
          aria-orientation="horizontal"
          aria-valuemax={44}
          aria-valuemin={0}
          aria-valuenow={0}
          className="MuiSlider-thumb MuiSlider-thumbColorPrimary PrivateValueLabel-open-2 PrivateValueLabel-thumb-1"
          data-index={0}
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseLeave={[Function]}
          onMouseOver={[Function]}
          role="slider"
          style={
            Object {
              "left": "0%",
            }
          }
          tabIndex={0}
        >
          <span
            className="PrivateValueLabel-offset-3 MuiSlider-valueLabel"
          >
            <span
              className="PrivateValueLabel-circle-4"
            >
              <span
                className="PrivateValueLabel-label-5"
              >
                11/05/2020
              </span>
            </span>
          </span>
        </span>
      </span>
    </article>
    <article>
      <ul>
        <li>
          mai/2020
        </li>
        <li>
          out/2020
        </li>
        <li>
          mar/2021
        </li>
        <li>
          jul/2021
        </li>
        <li>
          jan/2022
        </li>
      </ul>
    </article>
  </section>
  <section>
    <article>
      <label
        htmlFor="methodSelector"
      >
        Modo de visualização:
      </label>
      <div
        className="MuiInputBase-root MuiInput-root MuiInput-underline"
        data-testid="methodSelector"
        onClick={[Function]}
      >
        <div
          aria-haspopup="listbox"
          aria-labelledby="methodSelector"
          className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
          id="methodSelector"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          role="button"
          tabIndex={0}
        >
          Parcial
        </div>
        <input
          aria-hidden={true}
          className="MuiSelect-nativeInput"
          onAnimationStart={[Function]}
          onChange={[Function]}
          tabIndex={-1}
          value="partial"
        />
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSelect-icon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
    </article>
    <article>
      <label
        htmlFor="playButton"
      >
        Simular passagem de tempo:
      </label>
      <button
        className="styles__Button-sc-1n80o5l-1 ksEoZe"
        data-testid="playButton"
        id="playButton"
        onClick={[Function]}
        type="button"
      >
        Play
      </button>
    </article>
  </section>
</form>
`;
